examples/Efficient Cavity Control with SNAP Gates.py

```diff
@@ -0,0 +1,231 @@
+                  blocks on the vacuum state
+    """
+    displace = Displace(N)
+    for t in range(T):
+        blk = jnp.dot(displace(params[0 + t]), initial)
+        blk = jnp.dot(snap(hilbert_size, params[T + (t * hilbert_size) : T + (t + 1) * hilbert_size]), blk)
+        initial = jnp.dot(displace(-params[0 + t]), blk) # displace(alpha)^{\dagger} = displace(-alpha)
+
+    evolved = initial
+    return evolved
+
+
+# In[4]:
+
+
+def cost(params, initial, target, T, hilbert_size):
+    """
+    Calculates the cost between the target state and 
+    the one evolved by the action of three blocks.
+    
+    Args:
+    -----
+        params (jnp.array): alpha and theta params of Displace and SNAP respectively
+        initial (jnp.array): initial state to apply the blocks on
+        target (jnp.array): desired state
+    
+    Returns:
+    --------
+        cost (float): cost at a particular parameter vector
+    """
+    evo = add_blocks(initial, T, hilbert_size, params)
+    return 1 - fidelity(target, evo)[0][0]
+
+
+# # Initialize parameters
+# 
+# We start with an alomst alternating sequence of $\alpha$ and $\theta$ parameters as a gopod starting starting point, rather than random guessing. The target state is a binomial state
+# 
+# \begin{equation}
+# b_{1} = \frac{\sqrt 3 |3> + |9>}{2}
+# \end{equation}
+
+# In[5]:
+
+
+T = 3 # Number of blocks, B, to apply
+N = 10 # dimension of Fock space
+
+alphas = jnp.array([1., 0.5, 1.]) # Displace parameters
+theta1, theta2, theta3 = [0.5], [0.5, 1.5, 0.5], [0.5, 1.5, 0.5, 1.3] # SNAP parameters
+# NOTE: No input values to JAX differentiable functions should be int
+
+theta1, theta2, theta3 = pad_thetas(N, theta1), pad_thetas(N, theta2), pad_thetas(N, theta3)
+
+initial = basis(N, 0) # initial vacuum state
+target = (jnp.sqrt(3) * basis(10, 3) +  basis(10, 9)) / 2.0 # target state b1 as shown above|
+
+
+# # Gradient Descent -- Case in point for `qgrad`
+# 
+# This is where the power of `qgrad` comes in. Since qgrad's functions used in this notebook -- basis, to_dm, dag, Displace, fidelity -- support JAX, we can evaluate the gradient of the cost function in one line using JAX's `grad`. This saves us painstakingly evaluating the derivative of the cost function analytically.
+# 
+
+# In[6]:
+
+
+epochs = 17
+lr = 0.1 # learning rate
+max_iters = 100
+iters = 0
+params = jnp.concatenate((alphas, theta1, theta2, theta3)).reshape(3 * N + 3, 1)
+der_cost = grad(cost) # autodiff of the cost function
+
+params_hist = [params]
+
+for epoch in range (epochs):
+    iters = 0
+    while iters  < max_iters:
+        prev_params = params
+        der = der_cost(prev_params, initial, target, T, N)
+        params = prev_params - lr * der 
+        iters += 1
+    cost_val = cost(params, initial, target, T, N)
+    progress = [epoch+1, cost_val]
+    if ((epoch + 2) % 2 != 0):
+        params_hist.append(params)
+        print("Epoch: {:2f} | Cost: {:3f}".format(*jnp.asarray(progress)))
+
```

Comment:
The sub iterations iters < max_iters is confusion. We discussed this before too that what this does is simply do epochs * max_iters iteration. You can remove this and just increase the number of epochs as


```
params_hist = [(alphas, thetas),]
epochs = 100
lr = .1 # learning rate

for epoch in range (epochs):
    prev_thetas = thetas
    prev_alphas = alphas
    
    der = der_cost(prev_alphas, prev_thetas, initial_state, target_state)
    
    alphas = prev_alphas - lr * der[0]
    thetas = prev_thetas - lr * der[1]

    cost_val = cost(alphas, thetas, initial_state, target_state)
    progress = [epoch+1, cost_val]

    if ((epoch) % 10 != 0):
        params_hist.append((alphas, thetas))
        print("Epoch: {:2f} | Cost: {:3f}".format(*jnp.asarray(progress)))
```